File: examples/typescript/src/chat-server.ts
import { createReadStream, existsSync } from "node:fs";
import { createServer, type ServerResponse } from "node:http";
import { extname, resolve, sep } from "node:path";

import { Runner } from "marona";

import { buildAgent } from "./basic-agent.js";

const contentTypes: Record<string, string> = {
  ".css": "text/css; charset=utf-8",
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".svg": "image/svg+xml",
};

function json(response: ServerResponse, status: number, body: unknown): void {
  response.writeHead(status, { "content-type": "application/json; charset=utf-8" });
  response.end(JSON.stringify(body));
}

export async function runChatServer(): Promise<void> {
  const { agent } = await buildAgent();
  const port = Number.parseInt(process.env.PORT || "8080", 10);
  const uiDirectory = resolve(process.env.CHAT_UI_DIR || "apps/chat-ui");
  const server = createServer(async (request, response) => {
    if (request.method === "GET" && request.url === "/health") {
      json(response, 200, { status: "ok" });
      return;
    }
    if (request.method === "POST" && request.url === "/api/chat") {
      try {
        const chunks: Buffer[] = [];
        let size = 0;
        for await (const chunk of request) {
          const buffer = Buffer.from(chunk);
          size += buffer.length;
          if (size > 32_768) throw new Error("Request body exceeds 32 KiB.");
          chunks.push(buffer);
        }
        const payload = JSON.parse(Buffer.concat(chunks).toString("utf8")) as {
          message?: unknown;
          userId?: unknown;
          sessionId?: unknown;
        };
        const message = String(payload.message || "").trim();
        if (!message) throw new Error("message is required");
        const result = await Runner.run(agent, message, {
          userId: String(payload.userId || "example-user"),
          sessionId: String(payload.sessionId || "chat-ui-session"),
        });
        json(response, 200, { output: result.output });
      } catch (error) {
        const message = error instanceof Error ? error.message : "Invalid request.";
        json(response, 400, { error: message });
      }
      return;
    }
    if (request.method !== "GET") {
      json(response, 404, { error: "not_found" });
      return;
    }
    const requested = request.url === "/" ? "index.html" : decodeURIComponent(request.url || "").slice(1);
    const candidate = resolve(uiDirectory, requested);
    if (!candidate.startsWith(`${uiDirectory}${sep}`) || !existsSync(candidate)) {
      json(response, 404, { error: "not_found" });
      return;
    }
    response.writeHead(200, {
      "content-type": contentTypes[extname(candidate)] || "application/octet-stream",
    });
    createReadStream(candidate).pipe(response);
  });
  server.listen(port, "0.0.0.0", () => {
    console.log(`Marona Chat UI available at http://127.0.0.1:${port}`);
  });
}

File: examples/java/src/main/java/ai/marona/examples/ChatServer.java
package ai.marona.examples;

import ai.marona.Agent;
import ai.marona.Input;
import ai.marona.Marona;
import ai.marona.RunResult;
import ai.marona.Runner;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;
import java.util.concurrent.Executors;

public final class ChatServer {
    private static final int MAX_REQUEST_BYTES = 32_768;
    private static final ObjectMapper JSON = new ObjectMapper();
    private static final Map<String, String> CONTENT_TYPES = Map.of(
            ".css", "text/css; charset=utf-8",
            ".html", "text/html; charset=utf-8",
            ".js", "text/javascript; charset=utf-8",
            ".svg", "image/svg+xml"
    );

    private ChatServer() {
    }

    public static void run() {
        try {
            Marona marona = new Marona(Configuration.requiredApiKey());
            Agent agent = BasicAgentExample.buildAgent(marona);
            int port = Integer.parseInt(System.getenv().getOrDefault("PORT", "8080"));
            Path uiDirectory = Path.of(
                    System.getenv().getOrDefault("CHAT_UI_DIR", "../../apps/chat-ui")
            ).toAbsolutePath().normalize();
            HttpServer server = HttpServer.create(new InetSocketAddress("0.0.0.0", port), 0);
            server.createContext("/health", exchange -> json(
                    exchange,
                    200,
                    JSON.createObjectNode().put("status", "ok")
            ));
            server.createContext("/api/chat", exchange -> chat(exchange, agent));
            server.createContext("/", exchange -> staticFile(exchange, uiDirectory));
            server.setExecutor(Executors.newCachedThreadPool());
            Runtime.getRuntime().addShutdownHook(new Thread(() -> {
                server.stop(1);
                marona.close();
            }));
            server.start();
            System.out.println("Marona Chat UI available at http://127.0.0.1:" + port);
        } catch (IOException error) {
            throw new IllegalStateException("Could not start the Chat UI server.", error);
        }
    }

    private static void chat(HttpExchange exchange, Agent agent) throws IOException {
        if (!"POST".equals(exchange.getRequestMethod())) {
            json(exchange, 404, JSON.createObjectNode().put("error", "not_found"));
            return;
        }
        try {
            byte[] body = exchange.getRequestBody().readNBytes(MAX_REQUEST_BYTES + 1);
            if (body.length > MAX_REQUEST_BYTES) {
                throw new IllegalArgumentException("Request body exceeds 32 KiB.");
            }
            JsonNode payload = JSON.readTree(body);
            String message = payload.path("message").asText("").trim();
            if (message.isEmpty()) {
                throw new IllegalArgumentException("message is required");
            }
            String userId = payload.path("userId").asText("example-user");
            String sessionId = payload.path("sessionId").asText("chat-ui-session");
            RunResult result = Runner.run(agent, Input.text(message), userId, sessionId);
            ObjectNode response = JSON.createObjectNode().set("output", result.output());
            json(exchange, 200, response);
        } catch (IllegalArgumentException error) {
            json(exchange, 400, JSON.createObjectNode().put("error", error.getMessage()));
        } catch (RuntimeException error) {
            json(exchange, 500, JSON.createObjectNode().put("error", "request_failed"));
        }
    }

    private static void staticFile(HttpExchange exchange, Path uiDirectory) throws IOException {
        if (!"GET".equals(exchange.getRequestMethod())) {
            json(exchange, 404, JSON.createObjectNode().put("error", "not_found"));
            return;
        }
        String requested = "/".equals(exchange.getRequestURI().getPath())
                ? "index.html"
                : exchange.getRequestURI().getPath().substring(1);
        Path candidate = uiDirectory.resolve(requested).normalize();
        if (!candidate.startsWith(uiDirectory) || !Files.isRegularFile(candidate)) {
            json(exchange, 404, JSON.createObjectNode().put("error", "not_found"));
            return;
        }
        byte[] body = Files.readAllBytes(candidate);
        String filename = candidate.getFileName().toString();
        String extension = filename.contains(".")
                ? filename.substring(filename.lastIndexOf('.'))
                : "";
        exchange.getResponseHeaders().set(
                "Content-Type",
                CONTENT_TYPES.getOrDefault(extension, "application/octet-stream")
        );
        exchange.sendResponseHeaders(200, body.length);
        exchange.getResponseBody().write(body);
        exchange.close();
    }

    private static void json(HttpExchange exchange, int status, JsonNode payload)
            throws IOException {
        byte[] body = JSON.writeValueAsString(payload).getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().set("Content-Type", "application/json; charset=utf-8");
        exchange.sendResponseHeaders(status, body.length);
        exchange.getResponseBody().write(body);
        exchange.close();
    }
}
